File: 04_sorting/04_05_quick_sort/quick_sort.js
const partition = (arr = [], low, high) => {
    let i = low + 1;
    let j = high;

    while (true) {
        while (arr[i] < arr[low]) {
            i++;
            if (i >= high) {
                break;
            }
        }
        while (arr[j] > arr[low]) {
            j--;
            if (j <= low) {
                break;
            }
        }
        if (i >= j) {
            break;
        }
        [arr[i], arr[j]] = [arr[j], arr[i]]
    }

    
    [arr[low], arr[j]] = [arr[j], arr[low]]

    return j;
}

const quickSort = (arr = [], low, high) => {
    if (low >= high) {
        return;
    }

    const j = partition(arr, low, high);
    quickSort(arr, low, j - 1);
    quickSort(arr, j + 1, high);
}

const array = [3, 2, 8, 5, 1, 4, 23]

console.log("Before:-");
console.log(array);
quickSort(array, 0, array.length - 1)
console.log("After:-");
console.log(array);

File: 04_sorting/04_05_quick_sort/QuickSort.java
public class QuickSort {
    public static void main(String[] args) {
        int arr[] = { 13, 46, 24, 52, 20, 9 };
        System.out.println("Before:-");
        print(arr);
        quickSort(arr, 0, arr.length - 1);
        System.out.println();
        System.out.println("After:-");
        print(arr);
    }

    private static void quickSort(int arr[], int low, int high) {
        if (low >= high) {
            return;
        }

        int j = partition(arr, low, high);
        quickSort(arr, low, j - 1);
        quickSort(arr, j + 1, high);
    }

    private static int partition(int arr[], int low, int high) {
        int i = low + 1;
        int j = high;

        while (true) {
            while (arr[i] < arr[low]) {
                i++;
                if (i >= high) {
                    break;
                }
            }
            while (arr[j] > arr[low]) {
                j--;
                if (j <= low) {
                    break;
                }
            }
            if (i >= j) {
                break;
            }

            swap(arr, i, j);
        }

        swap(arr, low, j);
        return j;
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    private static void print(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
